kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/inband/RecordEncryptor.java

```diff
@@ -0,0 +1,149 @@
+        this.parcelBuffer = parcelBuffer;
+        this.wrapperBuffer = wrapperBuffer;
+        this.encryptionHeader = new Header[]{ new RecordHeader(ENCRYPTION_HEADER_NAME, new byte[]{ encryptionVersion.code() }) };
+    }
+
+    @Override
+    public void init(@NonNull Record kafkaRecord) {
+        if (encryptionScheme.recordFields().contains(RecordField.RECORD_HEADER_VALUES)
+                && kafkaRecord.headers().length > 0
+                && !kafkaRecord.hasValue()) {
+            // todo implement header encryption preserving null record-values
+            throw new IllegalStateException("encrypting headers prohibited when original record value null, we must preserve the null for tombstoning");
+        }
+        ByteBuffer transformedValue;
+        if (kafkaRecord.hasValue()) {
+            Parcel.writeParcel(encryptionVersion.parcelVersion(), encryptionScheme.recordFields(), kafkaRecord, parcelBuffer);
+            parcelBuffer.flip();
+            this.transformedValue = writeWrapper(parcelBuffer);
+            parcelBuffer.rewind();
+        }
+        else {
+            this.transformedValue = null;
+        }
+
+        Header[] oldHeaders = kafkaRecord.headers();
+        if (encryptionScheme.recordFields().contains(RecordField.RECORD_HEADER_VALUES) || oldHeaders.length == 0) {
+            transformedHeaders = encryptionHeader;
+        }
+        else {
+            transformedHeaders = new Header[1 + oldHeaders.length];
+            transformedHeaders[0] = encryptionHeader[0];
+            System.arraycopy(oldHeaders, 0, transformedHeaders, 1, oldHeaders.length);
+        }
+    }
+
+    @Nullable
+    private ByteBuffer writeWrapper(ByteBuffer parcelBuffer) {
+        switch (encryptionVersion.wrapperVersion()) {
+            case V1 -> {
+                var edek = keyContext.serializedEdek();
+                ByteUtils.writeUnsignedVarint(edek.length, wrapperBuffer);
+                wrapperBuffer.put(edek);
+                wrapperBuffer.put(AadSpec.NONE.code()); // aadCode
+                wrapperBuffer.put(CipherCode.AES_GCM_96_128.code());
+                keyContext.encodedSize(parcelBuffer.limit());
+                ByteBuffer aad = ByteUtils.EMPTY_BUF; // TODO pass the AAD to encode
+                keyContext.encode(parcelBuffer, wrapperBuffer); // iv and ciphertext
+            }
+        }
+        wrapperBuffer.flip();
+        return wrapperBuffer;
+    }
+
+    @Override
+    public void resetAfterTransform(Record record) {
+        wrapperBuffer.rewind();
+    }
+
+    @Override
+    public long transformOffset(Record record) {
+        return record.offset();
+    }
+
+    @Override
+    public long transformTimestamp(Record record) {
+        return record.timestamp();
+    }
+
+    @Override
+    public ByteBuffer transformKey(Record record) {
+        return record.key();
+    }
+
+    @Override
+    public ByteBuffer transformValue(Record kafkaRecord) {
+        return transformedValue;
```

Comment:
`transformedValue` is referencing the `wrapperBuffer` if we did the encryption. Then on `resetAfterTransform` we rewind the `wrapperBuffer`. I see we have the javadoc 
```
{@code RecordTransform}s are stateful and retain ownership of the {@link ByteBuffer}s returned
 *    by their {@code transform*()} methods.</p>
 ```
 should we doc more explicitly that the buffers contents should not be relied on after calling `resetAfterTransform`?

Reply:
Where would you document that?

Given the complexities we ended up with trying to be sure to close `BAMRB`s in the Collectors, I'm inclined to wrap up the `Stream` code (e.g. expose `io.kroxylicious.filter.encryption.records.MemoryRecordsUtils#concat`, rather than `io.kroxylicious.filter.encryption.records.MemoryRecordsUtils#concatCollector`). Doing that makes would mean there are fewer places to get this wrong (on the caller side).

Reply:
do we need to rewind the buffer on reset? It looks like it would be a no-op with the flip during `init`.

Reply:
oh, is it to accommodate users reading from it

Reply:
> oh, is it to accommodate clients reading from it between init and reset.

We could do that using `return transformedValue.asReadOnlyBuffer()`, that way clients get their own `position`. But we've given out access to this buffer which will also be used for the next record, and the next. So read-only buffer isn't enough of a guarantee. We need the caller to also do their thing with the buffer before they move on to the next record. Which is why we have this reset method, as a way for the caller to say "OK, I'm done with this one".